src/application-navigation/private/workspace-layout/WorkspaceLayout.jsx

```diff
@@ -0,0 +1,447 @@
+  } else {
+    initialWorkspaceSize = initialSizeForBreakpoint(activeBreakpoint);
+  }
+
+  const [workspaceSize, setWorkspaceSize] = React.useState(initialWorkspaceSize);
+  const hasOverlayWorkspace = activeBreakpoint === 'tiny' || activeBreakpoint === 'small' || workspaceSize.type === 'overlay';
+  const isLargeFormFactor = activeBreakpoint === 'large' || activeBreakpoint === 'huge' || activeBreakpoint === 'enormous';
+  const [workspaceIsVisible, setWorkspaceIsVisible] = React.useState(!hasOverlayWorkspace && workspace && workspace.props.initialIsOpen);
+  const hasWorkspace = !!workspace;
+
+  const actionsContextValue = React.useMemo(() => {
+    let actions = [];
+    let actionLabel;
+    if (workspaceIsVisible) {
+      actionLabel = applicationIntl.formatMessage({ id: 'terraApplication.workspace.showWorkspaceLabel' });
+    } else {
+      actionLabel = applicationIntl.formatMessage({ id: 'terraApplication.workspace.hideWorkspaceLabel' });
+    }
+
+    if (hasWorkspace) {
+      actions = [...actions, {
+        key: 'workspace-layout-toggle-workspace-panel',
+        label: actionLabel,
+        icon: workspaceIsVisible ? <IconPanelRight /> : <IconPanelLeft />,
+        onSelect: () => {
+          setWorkspaceIsVisible(state => !state);
+        },
+      }];
+    }
+
+    return ({
+      actions,
+    });
+  }, [hasWorkspace, workspaceIsVisible, applicationIntl]);
+
+  React.useEffect(() => {
+    if (!lastActiveSizeRef.current) {
+      lastActiveSizeRef.current = activeBreakpoint;
+      return;
+    }
+
+    if (lastActiveSizeRef.current === activeBreakpoint) {
+      return;
+    }
+
+    lastActiveSizeRef.current = activeBreakpoint;
+
+    if (activeBreakpoint === 'tiny' || activeBreakpoint === 'small') {
+      setWorkspaceSize({
+        scale: undefined,
+        type: undefined,
+      });
+    } else if (activeBreakpoint === 'medium') {
+      const scale = userSelectedScaleRef.current || workspaceSize.scale;
+
+      if (scale === undefined || scale <= 0.5) {
+        setWorkspaceSize({
+          scale: undefined,
+          type: 'split',
+        });
+      } else if (scale > 0.5) {
+        setWorkspaceSize({
+          scale: undefined,
+          type: 'overlay',
+        });
+      }
+    } else if (activeBreakpoint === 'large' || activeBreakpoint === 'huge' || activeBreakpoint === 'enormous') {
+      setWorkspaceSize({
+        scale: userSelectedScaleRef.current || 0,
+        type: undefined,
+      });
+    }
+  }, [workspaceSize, activeBreakpoint]);
+
+  const lastWorkspaceOpenState = React.useRef(workspaceIsVisible);
+  React.useEffect(() => {
+    if (workspaceIsVisible && !lastWorkspaceOpenState.current && workspacePanelRef.current) {
+      setTimeout(() => { workspacePanelRef.current.focus(); }, 0);
+    } else if (!workspaceIsVisible && lastWorkspaceOpenState.current) {
+      setTimeout(() => {
+        // TODO: evaluate flexing focus element based on overlay state or previous active element
```

Comment:
Is this TODO still needed now or something to do in the future?

Reply:
Current functionality is desired, I'll remove the TODO.